### PR TITLE
Implement HttpContext against Core HttpContext

### PR DESCRIPTION
**Pull request by twsouthwick**
This change starts plumbing ASP.NET Core HttpContext through the HttpContext
and sets up the caching mechanism used to ensure only a single instance
of the HttpContext is created per context.

Conversion between the adapters and core types are provided via both
implicit operators as well as extension methods that get or unwrap the
adapters. The HttpContext adapter is stored in the core
HttpContext.Features so that only a single instance is used for each
instance of the core context. The additional members (such as .Request
or .Response) are cached within the adapters and only created on demand.

This change also includes the HttpContextWrapper, HttpRequestWrapper,
and HttpResponseWrapper which are converted and stored in a similar way
as to the HttpContext adapter.

### CODE REVIEW

**Comment by twsouthwick** [comment]
/cc @davidfowl  @pranavkm 

I'm especially interested in your thoughts of storing the adapter in `HttpContext.Features`. I could swap it to use `.Items`, but this seems like it would have a bit better perf for those who are using it often since `.Items` is built on the `.Features`.

File: src/SystemWebAdapters/src/SystemWebAdaptersExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+            if (result is null)
+            {
+                result = new HttpContextWrapper(context);
+                context.Features.Set(result);
+            }
+
+            return result!;
+        }
+
+        [return: NotNullIfNotNull("request")]
```
Comment (davidfowl) [inline src/SystemWebAdapters/src/SystemWebAdaptersExtensions.cs:51]: Why do we need these?

Comment (twsouthwick) [reply src/SystemWebAdapters/src/SystemWebAdaptersExtensions.cs:51]: That was to provide support for `adapter.UnwrapAdapter()` and `context.GetAdapter()` to have nullability information flow correctly if the objects calling it were null.

**Comment by davidfowl** [comment]
> I'm especially interested in your thoughts of storing the adapter in HttpContext.Features. I could swap it to use .Items, but this seems like it would have a bit better perf for those who are using it often since .Items is built on the .Features.

This is good (this was my idea as well).
FeelMeThinkAboutItGIF (2)

File: src/SystemWebAdapters/src/SystemWebAdaptersExtensions.cs (hunk 1 above)
Comment (davidfowl) [reply src/SystemWebAdapters/src/SystemWebAdaptersExtensions.cs:51]: But why do we need these public

Comment (twsouthwick) [reply src/SystemWebAdapters/src/SystemWebAdaptersExtensions.cs:51]: Pushed a change to move them internal

**Pull request merged**